backend/auth-service/src/main/kotlin/com/nosql/authservice/component/impl/DefaultUserComponent.kt

```diff
@@ -0,0 +1,109 @@
+class DefaultUserComponent(
+    private val userRepository: UserRepository,
+) : UserComponent {
+
+    private val logger: Logger by logger()
+
+    @Transactional
+    override suspend fun saveIfNotExists(user: UserEntity): UserEntity {
+        val operationDetails = "Save if not exists 'auth' record with login = "
+
+        logger.logBefore(operationDetails)
+
+        assertUserNotExists(user)
+
+        return userRepository.save(user)
+            .onErrorMap { handleError(it, operationDetails) }
+            .doOnSuccess { logger.logSuccess(operationDetails) }
+            .awaitSingle()
+    }
+
+    @Transactional(readOnly = true)
+    override suspend fun existsByLogin(login: String): Boolean {
+        val operationDetails = "Check if 'auth' record by login = '$login' exists"
+
+        logger.logBefore(operationDetails)
+
+        return userRepository.existsByLogin(login)
+            .onErrorMap { handleError(it, operationDetails) }
+            .doOnSuccess { logger.logSuccess(operationDetails) }
+            .awaitSingle()
+    }
+
+    @Transactional(readOnly = true)
+    override suspend fun getByLoginAndPasswordHash(userDto: UserDto): UserEntity {
+        val (login, passwordHash) = userDto
+        val operationDetails = "Get 'auth' record by login = '$login' and password_hash = '$passwordHash'"
+
+        logger.logBefore(operationDetails)
+
+        return userRepository.getByLoginAndPasswordHash(login, passwordHash)
+            .switchIfEmpty(
+                Mono.error(
+                    NotFoundException(
+                        description = "Not found 'auth' record by given login = '$login' and " +
+                                "password_hash = '$passwordHash'"
+                    ),
+                )
+            )
+            .onErrorMap { handleError(it, operationDetails) }
+            .doOnSuccess { logger.logSuccess(operationDetails) }
+            .awaitSingle()
+    }
+
+    private suspend fun assertUserNotExists(user: UserEntity) {
+        if (existsByLogin(user.login))
+            throw UserAlreadyExistsException("User with login = '${user.login}' already exists")
+    }
+
+    private fun handleError(
+        error: Throwable,
+        operationDetails: String,
+    ): ApplicationException {
+
+        logger.logFailed(operationDetails, error)
+
+        val operationFailedWithMessage = "$operationDetails: failed with"
+        val errorDetails = if (error is ApplicationException) error.description else "something went wrong"
+
+        return when (error) {
+            is DataIntegrityViolationException -> {
+                DataIntegrityViolationApplicationException()
+            }
+            is ApplicationException -> {
+                error.apply { description = """$operationFailedWithMessage error: $errorDetails""" }
+            }
+            else -> {
+                ApplicationException(
+                    code = ApplicationExceptionCodeEnum.UNEXPECTED_ERROR,
```

Comment:
I think you should make `UNEXPECTED_ERROR` as a default value for the `ApplicationException.code` field

Reply:
fixed